### ISSUE TITLE
Camera near clip value can't be negative in editor

### ISSUE THREAD

**Issue by dmlo** [godotengine/godot#36446]
**Godot version:**
3.2

**OS/device including version:**
Win 10 / Dell Precision 4700

**Issue description:**
I was unable to set the camera near clipping value to a negative number resulting in a visible clip problem. Negative values are prohibited by the inspector but are settable via .set_znear(-1000). I believe the editor should allow negative values as a negative near clip is desirable for setting up an isometric camera.

**Steps to reproduce:**
1. Add a camera to a scene
2. Set perspective to orthagonal
3. Attempt to enter a negative number in a camera's near clip field in the godot editor. The value mins at 0.01 instead of allowing a negative value

**Minimal reproduction project:**
not applicable

Thank you for your consideration of this change and thank you for a fantastic game engine.

**Comment by Calinou** [comment]
Is there even such a thing as a negative near clip value in computer graphics?

Depth buffer precision gets worse as you bring the near plane closer to 0. It would be infinitely bad if it was 0 :slightly_smiling_face:

**Comment by dmlo** [comment]
Yes... When setting up an isometric scene a zero clip clips all objects at the z-axis of the camera but because of the orthographic projection that line is visible in the viewport.

![image](https://user-images.githubusercontent.com/7753993/75094013-115b5200-5555-11ea-8576-7d9aa1b9e375.png)

**Comment by dmlo** [comment]
and with the camera set to
self.set_znear(-1000)

![image](https://user-images.githubusercontent.com/7753993/75094038-3ea80000-5555-11ea-88f1-d7d469d1e071.png)

**Comment by Zylann** [comment]
As a workaround you can pull back the camera, which has the same effect

**Comment by dmlo** [comment]
@Zylann Agreed, but I am filing a bug because the inspector does not permit values that the underlying method allows.

**Comment by Calinou** [comment]
@dmlo What does it look like if you set the near plane value below 0 in perspective mode? We can't set hints dynamically.

**Comment by dmlo** [comment]
No objects are visible in perspective with set_znear(-1).

**Comment by Calinou** [comment]
@dmlo Indeed, this is what I feared :slightly_smiling_face: 

I guess we could clamp the effective near plane value when using perspective mode.

**Comment by dmlo** [comment]
@Calinou I did a test in another game engine and that's exactly how they handle it. You can set a negative clip value in ortho mode but when you switch to perspective it clamps to 0.01,

I tried @Zylann's idea of pulling the camera back but since the ortho view is just a column it had no effect.

If not a negative near clip, are there any other methods for avoiding the clipping problem in orthographic projection?

**Comment by Zylann** [comment]
> I tried @Zylann's idea of pulling the camera back but since the ortho view is just a column it had no effect.

This gotta have an effect, really. It does in perspective, and if it doesnt in ortho then we have another problem...

**Comment by darksylinc** [comment]
I came across this ticket by accident while googling on negative near clip planes.

I only wanted to point out code may assume near clipping can't be negative, thus causing weird bugs.

Off the top off my head I can think of PSSM (Parallel Split Shadow Mapping) / CSM (Cascaded Shadow Mapping) although any algorithm using cascades really, where the code looks at the camera bounds to determine what are the visible receivers and thus position the shadow mapping camera at the optimal location to exclude unnecessary casters.

While it is possible to handle negative near planes (ortho) in PSSM/CSM, it needs be checked whether it's handled correctly. Same goes for other code (e.g. screen ray picking comes to mind).

Alternatively it may be better/safer to reposition the camera backwards when using negative near clip planes, which is the correct thing to do.

**Comment by Calinou** [comment]
> (on a side note, it would be nice if the camera's gizmo lines were showing the real clipping volume)

@Zylann How do you think this could be done? Drawing a line from Znear to Zfar may be too invasive for high Zfar values (the default is 4000).

**Comment by Zylann** [comment]
I just suggested this because I saw that's what Unity does.

**Comment by louiidev** [comment]
I'm happy to make a PR for this

**Comment by Calinou** [comment]
> I'm happy to make a PR for this

There's already an open PR for this: https://github.com/godotengine/godot/pull/54850
